From 7a4ce1007edcfe37dec3f8c27158f428dc28c3d8 Mon Sep 17 00:00:00 2001
From: Dirk Neukirchen <dirkneukirchen@web.de>
Date: Wed, 6 May 2015 07:54:30 +0200
Subject: [PATCH 03/16] grub2: fix musl build

adopt upstream patch that removes _llseek

error with musl:
grub-core/osdep/unix/hostdisk.c: In function 'grub_util_fd_seek':
grub-core/osdep/unix/hostdisk.c:89:14: error: invalid storage class for function '_llseek'
   static int _llseek (uint filedes, ulong hi, ulong lo,

Signed-off-by: Dirk Neukirchen <dirkneukirchen@web.de>
---
 .../220-upstream_require_long_fs_libc.patch        | 80 ++++++++++++++++++++++
 1 file changed, 80 insertions(+)
 create mode 100644 package/boot/grub2/patches/220-upstream_require_long_fs_libc.patch

diff --git a/package/boot/grub2/patches/220-upstream_require_long_fs_libc.patch b/package/boot/grub2/patches/220-upstream_require_long_fs_libc.patch
new file mode 100644
index 0000000..a147fd9
--- /dev/null
+++ b/package/boot/grub2/patches/220-upstream_require_long_fs_libc.patch
@@ -0,0 +1,80 @@
+From 3bac4caa2bc64db313aaee54fffb90383e118517 Mon Sep 17 00:00:00 2001
+From: Felix Janda <felix.janda@posteo.de>
+Date: Thu, 22 Jan 2015 18:54:36 +0000
+Subject: Remove direct _llseek code and require long filesystem libc.
+
+---
+--- a/INSTALL
++++ b/INSTALL
+@@ -35,6 +35,7 @@ configuring the GRUB.
+ * GNU binutils 2.9.1.0.23 or later
+ * Flex 2.5.35 or later
+ * Other standard GNU/Unix tools
++* a libc with large file support (e.g. glibc 2.1 or later)
+ 
+ On GNU/Linux, you also need:
+ 
+--- a/configure.ac
++++ b/configure.ac
+@@ -361,6 +361,14 @@ case "$host_os" in
+    ;;
+ esac
+ 
++case "$host_os" in
++  cygwin | windows* | mingw32* | aros*)
++     ;;
++  *)
++     AC_CHECK_SIZEOF(off_t)
++     test x"$ac_cv_sizeof_off_t" = x8 || AC_MSG_ERROR([Large file support is required]);;
++esac
++
+ if test x$USE_NLS = xno; then
+   HOST_CFLAGS="$HOST_CFLAGS -fno-builtin-gettext"
+ fi
+--- a/grub-core/osdep/unix/hostdisk.c
++++ b/grub-core/osdep/unix/hostdisk.c
+@@ -48,11 +48,6 @@
+ #ifdef __linux__
+ # include <sys/ioctl.h>         /* ioctl */
+ # include <sys/mount.h>
+-# if !defined(__GLIBC__) || \
+-        ((__GLIBC__ < 2) || ((__GLIBC__ == 2) && (__GLIBC_MINOR__ < 1)))
+-/* Maybe libc doesn't have large file support.  */
+-#  include <linux/unistd.h>     /* _llseek */
+-# endif /* (GLIBC < 2) || ((__GLIBC__ == 2) && (__GLIBC_MINOR < 1)) */
+ #endif /* __linux__ */
+ 
+ grub_uint64_t
+@@ -79,24 +74,6 @@ grub_util_get_fd_size (grub_util_fd_t fd
+   return st.st_size;
+ }
+ 
+-#if defined(__linux__) && (!defined(__GLIBC__) || \
+-        ((__GLIBC__ < 2) || ((__GLIBC__ == 2) && (__GLIBC_MINOR__ < 1))))
+-  /* Maybe libc doesn't have large file support.  */
+-int
+-grub_util_fd_seek (grub_util_fd_t fd, grub_uint64_t off)
+-{
+-  loff_t offset, result;
+-  static int _llseek (uint filedes, ulong hi, ulong lo,
+-		      loff_t *res, uint wh);
+-  _syscall5 (int, _llseek, uint, filedes, ulong, hi, ulong, lo,
+-	     loff_t *, res, uint, wh);
+-
+-  offset = (loff_t) off;
+-  if (_llseek (fd, offset >> 32, offset & 0xffffffff, &result, SEEK_SET))
+-    return -1;
+-  return GRUB_ERR_NONE;
+-}
+-#else
+ int
+ grub_util_fd_seek (grub_util_fd_t fd, grub_uint64_t off)
+ {
+@@ -107,7 +84,6 @@ grub_util_fd_seek (grub_util_fd_t fd, gr
+ 
+   return 0;
+ }
+-#endif
+ 
+ 
+ /* Read LEN bytes from FD in BUF. Return less than or equal to zero if an
-- 
2.4.0

